File: agentlang/ast.py
from __future__ import annotations

from dataclasses import dataclass, field
from types import MappingProxyType


@dataclass(frozen=True)
class Span:
    line: int
    col: int


@dataclass(frozen=True)
class TypeExpr:
    pass


@dataclass(frozen=True)
class PrimitiveType(TypeExpr):
    name: str


@dataclass(frozen=True)
class ListType(TypeExpr):
    item_type: TypeExpr


@dataclass(frozen=True)
class OptionType(TypeExpr):
    item_type: TypeExpr


@dataclass(frozen=True)
class ObjType(TypeExpr):
    fields: MappingProxyType[str, TypeExpr]

    def __init__(self, fields: dict[str, TypeExpr] | MappingProxyType[str, TypeExpr]) -> None:
        object.__setattr__(self, "fields", MappingProxyType(dict(fields)))


@dataclass(frozen=True)
class EnumType(TypeExpr):
    name: str


@dataclass(frozen=True)
class Param:
    name: str
    type_expr: TypeExpr


@dataclass(frozen=True)
class AgentDef:
    name: str
    model: str | None
    tools: tuple[str, ...]


@dataclass(frozen=True)
class TaskDef:
    name: str
    params: tuple[Param, ...]
    return_type: TypeExpr
    execution_mode: str = "handler"


@dataclass(frozen=True)
class ToolDef:
    name: str
    params: tuple[Param, ...]
    return_type: TypeExpr


@dataclass(frozen=True)
class TypeAliasDef:
    name: str
    type_expr: TypeExpr


@dataclass(frozen=True)
class EnumTypeDef:
    name: str
    variants: tuple[str, ...]


@dataclass(frozen=True)
class Expr:
    span: Span | None = field(default=None, compare=False, repr=False)


@dataclass(frozen=True)
class LiteralExpr(Expr):
    value: object = None


@dataclass(frozen=True)
class RefExpr(Expr):
    parts: tuple[str, ...] = ()


@dataclass(frozen=True)
class BinaryExpr(Expr):
    op: str = ""
    left: Expr = field(default_factory=Expr)
    right: Expr = field(default_factory=Expr)


@dataclass(frozen=True)
class ObjExpr(Expr):
    fields: MappingProxyType[str, Expr] = field(default_factory=lambda: MappingProxyType({}))

    def __init__(
        self,
        fields: dict[str, Expr] | MappingProxyType[str, Expr] | None = None,
        *,
        span: Span | None = None,
    ) -> None:
        object.__setattr__(self, "span", span)
        object.__setattr__(self, "fields", MappingProxyType(dict(fields or {})))


@dataclass(frozen=True)
class ListExpr(Expr):
    items: tuple[Expr, ...] = ()


@dataclass(frozen=True)
class Stmt:
    span: Span | None = field(default=None, compare=False, repr=False)


@dataclass(frozen=True)
class RunStmt(Stmt):
    target: str = ""
    task_name: str = ""
    args: MappingProxyType[str, Expr] = field(default_factory=lambda: MappingProxyType({}))
    agent_name: str | None = None
    retries: int = 0
    on_fail: str = "abort"
    fallback_expr: Expr | None = None
    timeout: float | None = None

    def __init__(
        self,
        target: str = "",
        task_name: str = "",
        args: dict[str, Expr] | MappingProxyType[str, Expr] | None = None,
        agent_name: str | None = None,
        retries: int = 0,
        on_fail: str = "abort",
        fallback_expr: Expr | None = None,
        timeout: float | None = None,
        *,
        span: Span | None = None,
    ) -> None:
        object.__setattr__(self, "span", span)
        object.__setattr__(self, "target", target)
        object.__setattr__(self, "task_name", task_name)
        object.__setattr__(self, "args", MappingProxyType(dict(args or {})))
        object.__setattr__(self, "agent_name", agent_name)
        object.__setattr__(self, "retries", retries)
        object.__setattr__(self, "on_fail", on_fail)
        object.__setattr__(self, "fallback_expr", fallback_expr)
        object.__setattr__(self, "timeout", timeout)


@dataclass(frozen=True)
class ParallelStmt(Stmt):
    branches: tuple[RunStmt, ...] = ()
    max_concurrency: int | None = None


@dataclass(frozen=True)
class IfStmt(Stmt):
    condition: Expr = field(default_factory=Expr)
    then_statements: tuple[Stmt, ...] = ()
    else_statements: tuple[Stmt, ...] | None = None


@dataclass(frozen=True)
class IfLetStmt(Stmt):
    binding: str = ""
    option_expr: Expr = field(default_factory=Expr)
    then_statements: tuple[Stmt, ...] = ()
    else_statements: tuple[Stmt, ...] | None = None


@dataclass(frozen=True)
class WhileStmt(Stmt):
    condition: Expr = field(default_factory=Expr)
    statements: tuple[Stmt, ...] = ()


@dataclass(frozen=True)
class BreakStmt(Stmt):
    pass


@dataclass(frozen=True)
class ContinueStmt(Stmt):
    pass


@dataclass(frozen=True)
class ReturnStmt(Stmt):
    expr: Expr = field(default_factory=Expr)


@dataclass(frozen=True)
class TryCatchStmt(Stmt):
    try_body: tuple[Stmt, ...] = ()
    error_var: str = ""
    catch_body: tuple[Stmt, ...] = ()


@dataclass(frozen=True)
class AssertStmt(Stmt):
    condition: Expr = field(default_factory=Expr)
    message: str | None = None


@dataclass(frozen=True)
class WorkflowStep:
    pass


@dataclass(frozen=True)
class WorkflowStageStep(WorkflowStep):
    target: str = ""
    agent_name: str = ""
    task_name: str = ""
    args: tuple[Expr, ...] = ()


@dataclass(frozen=True)
class WorkflowReviewStep(WorkflowStep):
    target: str = ""
    reviewer_agent: str = ""
    source: str = ""
    reviser_agent: str = ""
    revise_task_name: str = ""
    max_rounds: int = 0


@dataclass(frozen=True)
class WorkflowReturnStep(WorkflowStep):
    expr: Expr = field(default_factory=Expr)


@dataclass(frozen=True)
class PipelineDef:
    name: str = ""
    params: tuple[Param, ...] = ()
    return_type: TypeExpr = field(default_factory=TypeExpr)
    statements: tuple[Stmt, ...] = ()


@dataclass(frozen=True)
class WorkflowDef:
    name: str = ""
    params: tuple[Param, ...] = ()
    return_type: TypeExpr = field(default_factory=TypeExpr)
    steps: tuple[WorkflowStep, ...] = ()


@dataclass(frozen=True)
class TestBlockDef:
    name: str = ""
    statements: tuple[Stmt, ...] = ()


@dataclass(frozen=True)
class Program:
    agents: MappingProxyType[str, AgentDef] = field(default_factory=lambda: MappingProxyType({}))
    tools: MappingProxyType[str, ToolDef] = field(default_factory=lambda: MappingProxyType({}))
    tasks: MappingProxyType[str, TaskDef] = field(default_factory=lambda: MappingProxyType({}))
    pipelines: MappingProxyType[str, PipelineDef] = field(default_factory=lambda: MappingProxyType({}))
    workflows: MappingProxyType[str, WorkflowDef] = field(default_factory=lambda: MappingProxyType({}))
    type_aliases: MappingProxyType[str, TypeAliasDef] = field(default_factory=lambda: MappingProxyType({}))
    enum_types: MappingProxyType[str, EnumTypeDef] = field(default_factory=lambda: MappingProxyType({}))
    test_blocks: tuple[TestBlockDef, ...] = ()

    def __init__(
        self,
        agents: dict | MappingProxyType | None = None,
        tools: dict | MappingProxyType | None = None,
        tasks: dict | MappingProxyType | None = None,
        pipelines: dict | MappingProxyType | None = None,
        workflows: dict | MappingProxyType | None = None,
        type_aliases: dict | MappingProxyType | None = None,
        enum_types: dict | MappingProxyType | None = None,
        test_blocks: tuple[TestBlockDef, ...] | list[TestBlockDef] | None = None,
    ) -> None:
        object.__setattr__(self, "agents", MappingProxyType(dict(agents or {})))
        object.__setattr__(self, "tools", MappingProxyType(dict(tools or {})))
        object.__setattr__(self, "tasks", MappingProxyType(dict(tasks or {})))
        object.__setattr__(self, "pipelines", MappingProxyType(dict(pipelines or {})))
        object.__setattr__(self, "workflows", MappingProxyType(dict(workflows or {})))
        object.__setattr__(self, "type_aliases", MappingProxyType(dict(type_aliases or {})))
        object.__setattr__(self, "enum_types", MappingProxyType(dict(enum_types or {})))
        object.__setattr__(self, "test_blocks", tuple(test_blocks or ()))

File: src/ast.rs
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, BTreeSet};

pub const CURRENT_LANGUAGE_VERSION: &str = "0.5";
pub const SUPPORTED_LANGUAGE_VERSIONS: &[&str] = &["0.2", "0.3", "0.4", "0.5"];

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Span {
    pub line: usize,
    pub col: usize,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum TypeExpr {
    String,
    Number,
    Bool,
    Failure,
    List(Box<TypeExpr>),
    Option(Box<TypeExpr>),
    Result(Box<TypeExpr>, Box<TypeExpr>),
    Obj(BTreeMap<String, TypeExpr>),
    Record(String),
    Union(String),
    Enum(String),
    Alias(String),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Param {
    pub name: String,
    pub ty: TypeExpr,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Expr {
    Literal {
        value: serde_json::Value,
        span: Span,
    },
    Ref {
        parts: Vec<String>,
        span: Span,
    },
    Binary {
        op: BinaryOp,
        left: Box<Expr>,
        right: Box<Expr>,
        span: Span,
    },
    Obj {
        fields: BTreeMap<String, Expr>,
        span: Span,
    },
    Record {
        name: String,
        fields: BTreeMap<String, Expr>,
        span: Span,
    },
    Variant {
        type_name: String,
        variant: String,
        fields: BTreeMap<String, Expr>,
        span: Span,
    },
    Result {
        ok: bool,
        value: Box<Expr>,
        span: Span,
    },
    List {
        items: Vec<Expr>,
        span: Span,
    },
}
impl Expr {
    pub fn span(&self) -> Span {
        match self {
            Self::Literal { span, .. }
            | Self::Ref { span, .. }
            | Self::Binary { span, .. }
            | Self::Obj { span, .. }
            | Self::Record { span, .. }
            | Self::Variant { span, .. }
            | Self::Result { span, .. }
            | Self::List { span, .. } => *span,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum BinaryOp {
    Add,
    Eq,
    Ne,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RunStmt {
    pub target: String,
    pub callable: String,
    pub args: BTreeMap<String, Expr>,
    pub agent: Option<String>,
    pub retries: u32,
    pub retry_policy: RetryPolicy,
    pub retry_on: Vec<(String, String)>,
    pub on_fail: OnFail,
    pub timeout: Option<f64>,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RetryPolicy {
    pub initial_ms: u64,
    pub max_ms: u64,
    pub multiplier: f64,
    pub jitter: f64,
}

impl Default for RetryPolicy {
    fn default() -> Self {
        Self {
            initial_ms: 0,
            max_ms: 30_000,
            multiplier: 2.0,
            jitter: 0.0,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum OnFail {
    Abort,
    Use(Expr),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Stmt {
    Run(RunStmt),
    Approve {
        target: String,
        approval: String,
        prompt: String,
        expires_seconds: Option<u64>,
        delegate: Option<String>,
        span: Span,
    },
    ParallelMap {
        target: String,
        binding: String,
        items: Expr,
        run: RunStmt,
        max_concurrency: usize,
        failure_policy: FailurePolicy,
        span: Span,
    },
    Race {
        target: String,
        branches: Vec<RunStmt>,
        span: Span,
    },
    Parallel {
        branches: Vec<RunStmt>,
        max_concurrency: Option<usize>,
        span: Span,
    },
    If {
        condition: Expr,
        then_body: Vec<Stmt>,
        else_body: Vec<Stmt>,
        span: Span,
    },
    IfLet {
        binding: String,
        option: Expr,
        then_body: Vec<Stmt>,
        else_body: Vec<Stmt>,
        span: Span,
    },
    While {
        condition: Expr,
        body: Vec<Stmt>,
        span: Span,
    },
    Break(Span),
    Continue(Span),
    TryCatch {
        try_body: Vec<Stmt>,
        error_var: String,
        structured: bool,
        catch_body: Vec<Stmt>,
        span: Span,
    },
    Match {
        value: Expr,
        arms: Vec<MatchArm>,
        span: Span,
    },
    Assert {
        condition: Expr,
        message: Option<String>,
        span: Span,
    },
    Return {
        expr: Expr,
        span: Span,
    },
}
impl Stmt {
    pub fn span(&self) -> Span {
        match self {
            Self::Run(x) => x.span,
            Self::Approve { span, .. } => *span,
            Self::Parallel { span, .. }
            | Self::ParallelMap { span, .. }
            | Self::Race { span, .. }
            | Self::If { span, .. }
            | Self::IfLet { span, .. }
            | Self::While { span, .. }
            | Self::TryCatch { span, .. }
            | Self::Match { span, .. }
            | Self::Assert { span, .. }
            | Self::Return { span, .. }
            | Self::Break(span)
            | Self::Continue(span) => *span,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum FailurePolicy {
    FailFast,
    CollectAll,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MatchArm {
    pub type_name: String,
    pub variant: String,
    pub bindings: Vec<String>,
    pub body: Vec<Stmt>,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AgentDef {
    pub name: String,
    pub model: Option<String>,
    pub tools: Vec<String>,
    pub requirements: AgentRequirements,
    pub deployment: Option<AgentDeployment>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct AgentRequirements {
    pub capabilities: BTreeSet<String>,
    pub min_context: Option<u64>,
    pub max_latency_ms: Option<u64>,
    pub quality: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AgentDeployment {
    pub provider: String,
    pub model: String,
    #[serde(default)]
    pub endpoint: Option<String>,
    #[serde(default)]
    pub reasoning_effort: Option<String>,
    #[serde(default)]
    pub capabilities: BTreeSet<String>,
    #[serde(default)]
    pub context_window: Option<u64>,
    #[serde(default)]
    pub expected_latency_ms: Option<u64>,
    #[serde(default)]
    pub quality: Option<String>,
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TaskDef {
    pub name: String,
    pub params: Vec<Param>,
    pub return_type: TypeExpr,
    pub agent_task: bool,
    pub effects: BTreeSet<String>,
    pub idempotency: Idempotency,
    pub concurrency_group: Option<String>,
    pub concurrency_limit: Option<u32>,
    pub rate_limit_per_second: Option<u32>,
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ToolDef {
    pub name: String,
    pub params: Vec<Param>,
    pub return_type: TypeExpr,
    pub effects: BTreeSet<String>,
    pub idempotency: Idempotency,
    pub concurrency_group: Option<String>,
    pub concurrency_limit: Option<u32>,
    pub rate_limit_per_second: Option<u32>,
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
pub enum Idempotency {
    Pure,
    Idempotent,
    KeyedBy(String),
    NonIdempotent,
    #[default]
    Unspecified,
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RecordDef {
    pub name: String,
    pub fields: BTreeMap<String, TypeExpr>,
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct UnionDef {
    pub name: String,
    pub variants: BTreeMap<String, BTreeMap<String, TypeExpr>>,
}
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PipelineDef {
    pub name: String,
    pub params: Vec<Param>,
    pub return_type: TypeExpr,
    pub effects: Option<BTreeSet<String>>,
    pub budget: Option<ResourceBudget>,
    pub statements: Vec<Stmt>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct ResourceBudget {
    pub time_ms: Option<u64>,
    pub tokens: Option<u64>,
    pub cost_usd: Option<f64>,
    pub tool_calls: Option<u64>,
    pub retries: Option<u64>,
    pub concurrency: Option<u64>,
}
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TestBlock {
    pub name: String,
    pub statements: Vec<Stmt>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EvalDef {
    pub name: String,
    pub pipeline: String,
    pub dataset: String,
    pub trials: u32,
    pub baseline: Option<String>,
    pub assert_schema: bool,
    pub assert_expected: bool,
    pub max_latency_ms: Option<u64>,
    pub max_cost_usd: Option<f64>,
    pub semantic_grader: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Program {
    pub language_version: String,
    pub agents: BTreeMap<String, AgentDef>,
    pub tools: BTreeMap<String, ToolDef>,
    pub tasks: BTreeMap<String, TaskDef>,
    pub pipelines: BTreeMap<String, PipelineDef>,
    pub aliases: BTreeMap<String, TypeExpr>,
    pub records: BTreeMap<String, RecordDef>,
    pub unions: BTreeMap<String, UnionDef>,
    pub enums: BTreeMap<String, Vec<String>>,
    pub tests: Vec<TestBlock>,
    pub evals: BTreeMap<String, EvalDef>,
}

impl Default for Program {
    fn default() -> Self {
        Self {
            language_version: CURRENT_LANGUAGE_VERSION.into(),
            agents: BTreeMap::new(),
            tools: BTreeMap::new(),
            tasks: BTreeMap::new(),
            pipelines: BTreeMap::new(),
            aliases: BTreeMap::new(),
            records: BTreeMap::new(),
            unions: BTreeMap::new(),
            enums: BTreeMap::new(),
            tests: Vec::new(),
            evals: BTreeMap::new(),
        }
    }
}
